### PR TITLE
Adding `wo*y` commands using clipboard instead of [val]

### PR DESCRIPTION
**Pull request by cr**
When working with streaming ciphers, being able to XOR memory regions inside radare is super nifty. This implementation uses the clipboard as XOR key which is repeated when it is shorter than blocksize.

### CODE REVIEW

**Comment by radare** [comment]
You can already do this with the `wox` command. not in the clipboard.. and there are several other "encryption" algorithms and checksums that can be applied to a buffer, so i would prefer not to have more dupplicated commands, unless we follow the same schema.

For example. having `yo` subcommands to mimic the `wo` ones. and share as much code as possible between them.

**Comment by cr** [comment]
> You can already do this with the wox command.

*wox* makes you specify the key on the command line which is not feasible with a 1K stream cipher. The main motivation for *yX* was to be able to XOR two memory regions.

My first instinct was to extend the *wox* command accordingly, but you end up introducing two awkward length parameters. Using the clipboard instead makes it elegant to simply yank a key from somewhere, then XOR it where needed.

**Comment by radare** [comment]
oh i get your point. Using the clipboard as a place to store the key. My suggestion would be to add an `y` suffix to the `wo.` commands and take the clipboard as key instead of parsing the argument.

`> woxy`

**Comment by cr** [comment]
An elegant patch just turned ugly, but I may have made it. ```make tests``` runs without errors, but if the integration test here fails again, I don't know what's causing it, because it seems I can't read its output here.

**Comment by cr** [comment]
I couldn't bring myself to rewrite the arg parsing for the ```woE``` command.

**Comment by cr** [comment]
I am getting second thoughts here: Wouldn't it be more sensible to fall back to clipboard when no ```[val]``` is given instead of introducing the ```wo*y``` class of commands?

**Comment by cr** [comment]
I have the version ready that falls back to clipboard when no ```[val]``` is given. Let me know if you prefer that, and I'll push it here first thing in the morning.

**Comment by Maijin** [comment]
Hey dat looks cool stuff, have you checked #4280 and #4254 as it is a bit related?

File: libr/core/cmd_write.c [hunk 1]
```diff
@@ -94,21 +94,22 @@ static void cmd_write_op (RCore *core, const char *input) {
...
-		"woe"," [from to] [step] [wsz=1]","..  create sequence",
 		"wox"," [val]","^=  xor  (f.ex: wox 0x90)",
 		"woo"," [val]","|=  or",
 		"woA"," [val]","&=  and",
-		"woR","","random bytes (alias for 'wr $b')",
 		"wor"," [val]", ">>= shift right",
 		"wol"," [val]","<<= shift left",
 		"wo2"," [val]","2=  2 byte endian swap",
 		"wo4"," [val]", "4=  4 byte endian swap",
+		"wo*y", "", "same as their counterparts above, but using clipboard instead of [val]",
```
Comment (radare) [inline libr/core/cmd_write.c]: this shuold be alphabetically sorted

File: libr/core/io.c [hunk 2]
```diff
@@ -97,23 +97,42 @@ R_API int r_core_dump(RCore *core, const char *file, ut64 addr, ut64 size, int a
...
+			}
+		} else len = 0;
+	} else {  // use clipboard as key
+		if (op!='e') {
+			len = core->yank_buf->length;
+			if (len <= 0) {
+				eprintf ("Clipboard is empty\n");
+				goto beach;
+			}
+			str = (char *)malloc (len);
```
Comment (radare) [inline libr/core/io.c]: you can use `str = r_mem_dup(core->yank_buf->buf, len);` to save some lines here

File: libr/core/io.c [hunk 3]
```diff
@@ -97,23 +97,42 @@ R_API int r_core_dump(RCore *core, const char *file, ut64 addr, ut64 size, int a
...
+	// fill key buffer either from arg or from clipboard
+	if (arg) {  // parse arg for key
+		str = (char *)malloc (strlen (arg)+1);
+		if (!str)
 			goto beach;
-		}
-	} else len = 0;
+		if (op!='e') {
+			len = r_hex_str2bin (arg, (ut8 *)str);
+			if (len < 0) {
```
Comment (radare) [inline libr/core/io.c]: if len == 0 is also an invalid length, right?

File: libr/core/io.c [hunk 4]
```diff
@@ -97,23 +97,42 @@ R_API int r_core_dump(RCore *core, const char *file, ut64 addr, ut64 size, int a
...
 		goto beach;
 	memcpy (buf, core->block, core->blocksize);
-	if (op!='e') {
-		len = r_hex_str2bin (arg, (ut8 *)str);
-		if (len < 0) {
-			eprintf ("Invalid hexpair string\n");
+
+	// fill key buffer either from arg or from clipboard
+	if (arg) {  // parse arg for key
+		str = (char *)malloc (strlen (arg)+1);
```
Comment (radare) [inline libr/core/io.c]: better use calloc to avoid uninitialized memory in case r_hex_str2bin doesnt fills up all the buffer

Comment (radare) [inline libr/core/io.c]: spaces around `+`

**Comment by radare** [comment]
Sorry for the late reply, its been a crazy day yesterday. Can you address the comments and update the PR? looks good to merge

**Comment by cr** [comment]
@radare, please comment on https://github.com/radare/radare2/pull/4533#issuecomment-206042631 . I much prefer that solution.

File: libr/core/io.c (hunk 3 above)
Comment (cr) [inline libr/core/io.c]: You tell me! ;) I didn't touch that part of the code. Will change to ```(len <= 0)``` in my amended commit.

File: libr/core/io.c (hunk 4 above)
Comment (cr) [inline libr/core/io.c]: OK, will fix these for you.

File: libr/core/cmd_write.c (hunk 1 above)
Comment (cr) [inline libr/core/cmd_write.c]: It's been a mess before, so it was hard to make out the right spot. I tried to work surgically. Shouldn't a cleanup here be part of a different pull request?

File: libr/core/io.c (hunk 2 above)
Comment (cr) [inline libr/core/io.c]: Sweet! Good to know, will use that instead.

**Comment by cr** [comment]
TBH, I have never gotten in that much trouble for not touching the existing code. I am thankful that you pointed out the mistakes I was about to leave in place, but I also feel that we might be better off to separate my feature work from what seems to be a much-needed audit/rewrite of the existing code.

File: libr/core/io.c (hunk 4 above)
Comment (cr) [inline libr/core/io.c]: > better use calloc to avoid uninitialized memory in case r_hex_str2bin doesnt fills up all the buffer

After giving it a closer look, zero-filling the buffer by using ```calloc()``` won't change anything but make it slower. What you're suggesting makes sense for string operations that might end up non-zero-terminated, but ```r_hex_str2bin()``` writes raw bytes (including zeroes) and returns the number of bytes it has written to the buffer. Buffer length is chosen such that it is *guaranteed to be much too large*, and all the code that works on ```str``` thereafter relies on ```len``` for termination.

If there's anything missing from this code, it is asserting that ```len <= strlen (arg) + 1``` to detect stack corruption, but given that *two nibble chars input result in one byte output*, one can safely omit that.

Btw, the ```+ 1``` turns out to be unnecessary. ```malloc(0)``` is fine to ```free()```, and ```len == 0``` will be caught right after.

File: libr/core/cmd_write.c (hunk 1 above)
Comment (radare) [inline libr/core/cmd_write.c]: r2 have a huge codebase and we try to have as much testcases as possible in order to not break stuff, because one of the main premises is the continuous refactoring, mainly because most of the stuff that is implemented in r2 is done "quickly" because is a feature someone needs for something NOW. but there are always better ways to do stuff, so that's where the refactoring enters the game. you can do this alphabetically sorting thing in the same PR, there's no real issue in sorting out a text array that will be processed by humans.

Comment (cr) [inline libr/core/cmd_write.c]: Thanks for the explanation. I'm sorry. I didn't mean to drive you into defense. :/

I will push along a separate commit in this pull request.

File: libr/core/io.c (hunk 4 above)
Comment (radare) [inline libr/core/io.c]: 👍

File: libr/core/cmd_write.c (hunk 1 above)
Comment (radare) [inline libr/core/cmd_write.c]: Into defense? Dont take me wrong. Im not trying to be offensive or anything. Just trying to give some feedback on every new contributor just to discuss better way to handle this.

Thats a very nice addition :) take a beer and chill a bit ;)

Ill merge after lunch

Thx

**Comment by cr** [comment]
Woah, someone merged this PR into master before we were done with the review. :(

**Comment by cr** [comment]
I am closing this PR. There's not much sense in trying to fix the broken merge on top of this.

**Comment by radare** [comment]
wtf?

you dont have to fix any merge and you can push -f in pr'd branches perfectly. github handles this

**Comment by radare** [comment]
ok moved to https://github.com/radare/radare2/pull/4548

**Comment by cr** [comment]
This PR was poisoned because I had force-pushed into this branch without realizing that it had already been merged.

File: libr/core/cmd_write.c (hunk 1 above)
Comment (cr) [inline libr/core/cmd_write.c]: No offense taken at all! I was just slightly worried that you felt offended because you started justifying radare code quality there.
